cylc/uiserver/tests/test_graphql.py

```diff
@@ -0,0 +1,161 @@
+def gql_query(jp_fetch):
+    """Perform a GraphQL request.
+
+    E.G:
+
+    response = await gql_query(
+        *('cylc', 'graphql'),
+        query='''
+            query {
+                workflows {
+                    id
+                    status
+                }
+            }
+        ''',
+    )
+
+    """
+
+    async def _fetch(*endpoint, query=None, headers=None):
+        headers = headers or {}
+        headers = {
+            **headers,
+            'Content-Type': 'application/json'
+        }
+        return await jp_fetch(
+            *endpoint,
+            method='POST',
+            headers=headers,
+            body=json.dumps({'query': query}, indent=4),
+        )
+
+    return _fetch
+
+
+async def test_query(gql_query, dummy_workflow):
+    """Test sending the most basic GraphQL query."""
+    # configure two dummy workflows so we have something to look at
+    await dummy_workflow('foo')
+    await dummy_workflow('bar')
+
+    # perform the request
+    response = await gql_query(
+        *('cylc', 'graphql'),
+        query='''
+            query {
+                workflows {
+                    id
+                    status
+                }
+            }
+        ''',
+    )
+    assert response.code == 200
+
+    # we should find the two dummy workflows in the response
+    body = json.loads(response.body)
+    assert body['data'] == {
+        'workflows': [
+            {
+                'id': 'me|foo',
+                'status': 'stopped',
+            },
+            {
+                'id': 'me|bar',
+                'status': 'stopped',
+            },
+        ]
+    }
+
+
+async def test_multi(gql_query, monkeypatch, cylc_uis, dummy_workflow):
+    """Ensure multi-mutation requests are properly forwarded.
+
+    * Multiple mutations can be specified in one request.
+    * These pass through the GraphQL routing where they are authorised
+      one by one.
+    * They should be issued to the scheduler one by one, the GraphQL
+      query must be altered so it only contains the single mutation.
+
+    """
+    # need at leat one workflow for the request to be forwarded
+    await dummy_workflow('foo')
+    await dummy_workflow('bar')
+
+    # log all calls that go via the multi_request interface
+    # (and prevent these calls from actually being made)
+    calls = []
+
+    async def _log(*args, **kwargs):
+        calls.append((args, kwargs))
+
+    monkeypatch.setattr(cylc_uis.workflows_mgr, 'multi_request', _log)
+
+    # issue three mutations
+    response = await gql_query(
+        *('cylc', 'graphql'),
+        query='''
+            mutation {
+                hold(workflows: ["*"], tasks: []) {
+                    result
+                }
+                pause(workflows: ["*"]) {
+                    result
+                }
+                stop(workflows: ["*"]) {
+                    result
+                }
+            }
+        ''',
+    )
+    assert response.code == 200
```

Comment:
consider 
```
if response.code != 200:
    raise SomePytestError(response.reason)

Reply:
If the test fails the full response (including the `response.reason`) and server logs are dumped in the pytest output so this is already well captured.